docker-compose-rule-core/src/main/java/com/palantir/docker/compose/execution/DockerCommandLocations.java

```diff
@@ -15,47 +15,53 @@
  */
 package com.palantir.docker.compose.execution;
 
-import static java.util.Arrays.asList;
-
 import java.io.File;
 import java.nio.file.Paths;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
 import java.util.Optional;
 import java.util.function.Predicate;
+import java.util.stream.Collectors;
 import java.util.stream.Stream;
-import org.apache.commons.lang3.StringUtils;
 
-public class DockerCommandLocations {
+import org.immutables.value.Value;
+
+@Value.Immutable
+public abstract class DockerCommandLocations {
     private static final Predicate<String> IS_NOT_NULL = path -> path != null;
     private static final Predicate<String> FILE_EXISTS = path -> new File(path).exists();
 
-    private final List<String> possiblePaths;
-    private final String exeName;
-    private final boolean lookInPath;
-
-    public DockerCommandLocations(String exeName, boolean lookInPath, String... possiblePaths) {
-        this.exeName = exeName;
-        this.lookInPath = lookInPath;
-        this.possiblePaths = asList(possiblePaths);
+    protected abstract String executableName();
+    protected abstract List<Optional<String>> additionalSearchLocations();
```

Comment:
agreed. however, having a signature of ```List<String>``` would fail in runtime (under certain circumstances). the ```Stream.of()``` solution is fine but the caller would have to know that they should call this method using this specific construct and not by using any other plain list.
putting the ```Optional``` in the signature ensures null-safety at compile time, right?